### PR TITLE
API module implementation

### PR DESCRIPTION
**Pull request by m-wojnar**
Dodałem implementację głównej klasy naszej biblioteki. Klasa pozwala na ustalenie typu agenta oraz środowiska, odczytanie observation space, a także inicjowanie oraz działanie na kilku instancjach danego agenta. W przyszłości planuję dodać jeszcze zapisywanie oraz odczytywanie z pliku stanu agenta, a także możliwość trenowania agenta.

### CODE REVIEW

File: reinforced_lib/rlib.py [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+            self, *,
+            agent_type: Callable = None,
+            agent_params: Dict = None,
+            env_type: type = None,
+            env_params: Dict = None,
+            log_type: Union[type, List[type]] = None,
+            log_params: Union[Dict, List[Dict]] = None,
+            no_env_mode: bool = False
+    ) -> None:
+        """
```
Comment (Wotaker) [inline reinforced_lib/rlib.py]: Zauważyłem że **sphinx** wymaga, aby docstringi były bezpośrednio pod nazwą klasy, a nie po `def __init__()`. Tylko wtedy można skorzystać z `.. autoclass::` generującego automatycznie dokumentacje klasy (analogicznie funkcji)

File: reinforced_lib/rlib.py [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+        if len(self._agents_states) > 0:
+            raise ForbiddenAgentChangeError()
+
+        agent_params = agent_params if agent_params else {}
+        self._agent = agent_type(**agent_params)
+
+        if not isinstance(self._agent, BaseAgent):
+            raise IncorrectAgentTypeError(agent_type)
+
+        if not self._no_env_mode and self._env:
```
Comment (Wotaker) [inline reinforced_lib/rlib.py:85]: Przeniósłbym to już albo do `set_env()`, albo do linijki 56, bo po prostu trochę z dupy to jest w tym `set_agent()`

File: reinforced_lib/rlib.py [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+            A number used to initialize the JAX pseudo-random number generator.
+
+        Returns
+        -------
+        out : int
+            The identifier of created instance.
+        """
+
+        agent_id = len(self._agents_states)
+        seed = seed if seed else 0
```
Comment (Wotaker) [inline reinforced_lib/rlib.py]: Gdzieś słyszałem (Może jeszcze na MOwNiT), że seed 0 nie jest zbyt dobrym pomysłem bo generowane liczby są słabo losowe. Nie wiem czy to prawda, ale no nie jest to problem zmienić np. na 1000, czy jakieś 42 😄

File: reinforced_lib/rlib.py [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+import jax.random
+
+from reinforced_lib.agents.base_agent import BaseAgent
+from reinforced_lib.envs.base_env import BaseEnv
+from reinforced_lib.utils.exceptions import *
+
+
+class RLib:
+    def __init__(
+            self, *,
```
Comment (Wotaker) [inline reinforced_lib/rlib.py:33]: Dlaczego zastosowałeś tutaj `*`? Bo rozumiem, że teraz wszyestkie parametry trzeba explicite wypisywać, to znaczy `agent_type=some_agent, agent_params={...}, ...`

File: reinforced_lib/rlib.py [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+            self.init()
+
+        self._agents_keys[agent_id], key = jax.random.split(self._agents_keys[agent_id])
+        state = self._agents_states[agent_id]
+
+        if not self._no_env_mode:
+            update_observations, sample_observations = self._env.transform(*args, **kwargs)
+
+        if isinstance(update_observations, dict):
+            state = self._agent(state, **update_observations)
```
Comment (Wotaker) [inline reinforced_lib/rlib.py:255]: czy nie zgubiłeś tutaj metody `update`? Tzn czy nie powinno być `state = self._agent.update(state, **update_observations)`, i analogicznie w `elif` i `else`?

File: reinforced_lib/rlib.py [hunk 6]
```diff
@@ -0,0 +1,337 @@
...
+        elif isinstance(update_observations, tuple):
+            state = self._agent(state, *update_observations)
+        else:
+            state = self._agent(state, update_observations)
+
+        if isinstance(sample_observations, dict):
+            state, action = self._agent(state, key, **sample_observations)
```
Comment (Wotaker) [inline reinforced_lib/rlib.py:262]: Tutaj pytanie analogiczne do poprzedniego, nie powinno być `state, action = self._agent.sample(state, key, **sample_observations)`

File: reinforced_lib/rlib.py [hunk 7]
```diff
@@ -0,0 +1,337 @@
...
+
+        Parameters
+        ----------
+        agent_id : int
+            The identifier of agent instance.
+        path : str
+            Path to the output file.
+        """
+
+        if not path:
```
Comment (Wotaker) [inline reinforced_lib/rlib.py:308]: Tutaj w przyszłości (jak już ogarniemy `fit()`) będzie można wyspecifikować jakąś domyślną `path`

**Review by Wotaker — CHANGES_REQUESTED** [review]
Wydaje mi się że mogłeś popełnic kilka błędów, dlatego nie daje narazie approve. Wszystko opisane w komentarzach

File: reinforced_lib/rlib.py (hunk 2 above)
Comment (m-wojnar) [reply reinforced_lib/rlib.py:85]: Nie wiem czy na pewno o to chodzi, ale linijki 85-86 są wprowadzone po to, że w wypadku, gdy użytkownik zmieni typ agenta, to musimy przeładować również środowisko (po to, aby sprawdzić czy mają kompatybilne observation space i wygenerować te funkcje transformacji observation space). Więc wydaje mi się, że to musi tutaj zostać, bo zmiana agenta mogłaby bez tego spowodować rozsypanie się komunikacji między agentem, a środowiskiem.

File: reinforced_lib/rlib.py (hunk 3 above)
Comment (m-wojnar) [reply reinforced_lib/rlib.py]: Jasne, też miałem wątpliwości czy 0 to dobry pomysł, ale pomyślałem, że to liczba jak każda inna i zostawiłem. Zmienię w takim razie na 42 ;)

File: reinforced_lib/rlib.py (hunk 4 above)
Comment (m-wojnar) [reply reinforced_lib/rlib.py:33]: Ponieważ parametry `agent_type`, `env_type`, `log_type`, a także `agent_params`, `env_params`, `log_params` są prawie identyczne i nie chciałem, aby użytkownik się pomylił wpisując na przykład agenta i środowisko w złej kolejności. Dlatego chciałem go "zmusić", aby explicite wpisał, co przekazuje. Ale jeśli uważasz, że to jest zły pomysł jak chodzi o UX, to mogę usunąć gwiazdkę.

File: reinforced_lib/rlib.py (hunk 5 above)
Comment (m-wojnar) [reply reinforced_lib/rlib.py:255]: Tak jest, zorientowałem się dopiero przy implementacji pull requesta #6 i tam to poprawiłem.

File: reinforced_lib/rlib.py (hunk 6 above)
Comment (m-wojnar) [reply reinforced_lib/rlib.py:262]: Jak wyżej

File: reinforced_lib/rlib.py (hunk 7 above)
Comment (m-wojnar) [reply reinforced_lib/rlib.py:308]: Też tak pomyślałem, ale na razie zostawiłem puste, bo to jest do ustalenia.

File: reinforced_lib/rlib.py (hunk 1 above)
Comment (m-wojnar) [reply reinforced_lib/rlib.py]: Jasne, rozumiem. Nie wiedziałem że tak to działa w **sphinx**. W takim razie przeniosę te ogólne komentarze we wszystkich klasach z init pod nazwę klasy.

File: reinforced_lib/rlib.py (hunk 4 above)
Comment (Wotaker) [reply reinforced_lib/rlib.py:33]: Nie no spoko, jest to zdecydowanie dobre uzasadnienie

File: reinforced_lib/rlib.py (hunk 1 above)
Comment (Wotaker) [reply reinforced_lib/rlib.py]: Super, daje resolved, trzeba po prostu o tym pamiętać

File: reinforced_lib/rlib.py (hunk 2 above)
Comment (Wotaker) [reply reinforced_lib/rlib.py:85]: Okey, rozumiem

File: reinforced_lib/rlib.py (hunk 3 above)
Comment (Wotaker) [reply reinforced_lib/rlib.py]: ok

File: reinforced_lib/rlib.py (hunk 5 above)
Comment (Wotaker) [reply reinforced_lib/rlib.py:255]: super

File: reinforced_lib/rlib.py (hunk 6 above)
Comment (Wotaker) [reply reinforced_lib/rlib.py:262]: ok

File: reinforced_lib/rlib.py (hunk 7 above)
Comment (Wotaker) [reply reinforced_lib/rlib.py:308]: To fajnie

**Review by Wotaker — APPROVED** [review]
Okey, to jak wszystkie konwersacje są resolved, daje approved
